File: web/services/citations.py
"""Turn SearchResult objects into a JSON-safe citation payload.

The search layer already computes everything a reader needs to judge an
answer — which document, which page, and how strongly it matched both
semantically and lexically — and the API used to throw all of it away. This
module is what carries it to the client.

Two things here are load-bearing:

1. **Type coercion.** ``ResultCombiner`` builds ``SearchResult`` from pandas
   cells and numpy index arrays, so ``page``/``chunk_id``/``score`` routinely
   arrive as ``np.int64`` / ``np.float32`` / ``NaN``. ``json.dumps`` raises
   ``TypeError: Object of type int64 is not JSON serializable`` on those, which
   surfaces as a bare 500. Every field is coerced, and the type hints on
   ``SearchResult`` are NOT trusted.

2. **Index alignment.** ``sources[i]["index"]`` must equal the ``[i]`` label the
   model saw in its prompt context, or a citation points at a card the model
   never read. Both are sliced with the same ``limit``.
"""

from __future__ import annotations

import logging
import math
import re
from collections.abc import Iterable
from dataclasses import dataclass
from typing import Any

logger = logging.getLogger(__name__)

SNIPPET_CHARS = 320

# Matches the prose citation format the model used before numbered context
# blocks: "[Source: Guidance.pdf, Page: 14]", page optional.
#
# `lead` captures at most ONE space before the citation, so a dropped citation
# takes its own spacing with it — "a claim [Source: X], and" becomes "a claim,
# and" — without any pass over the rest of the answer. Deliberately not `[ \t]*`:
# a greedy version would eat the indentation of a line that happens to start
# with a citation, and indentation is load-bearing markdown.
LEGACY_CITATION = re.compile(
    r"(?P<lead> ?)"
    r"\[\s*Source\s*:\s*(?P<doc>[^,\]]+?)\s*(?:,\s*Page\s*:\s*(?P<page>\d+)\s*)?\]",
    re.IGNORECASE,
)

# The server-side twin of CITE_SCAN in static/js/modules/citations.js.
#
# `[0-9]` rather than `\d`, and the difference is not cosmetic: Python's `\d`
# matches every Unicode decimal digit, JavaScript's (without the `u` flag)
# matches ASCII only. So "[١]" in Arabic-Indic digits matched here — int()
# parses it happily as 1 — while the browser saw no marker at all, and the
# answer gained a source with nothing on screen pointing at it. The Arabic
# language instruction asks the model to keep markers in Latin form, which is
# exactly the kind of rule this codebase does not rely on the model obeying.
#
# `{1,2}` is deliberate and must NOT be relaxed to `+`: "[123]" matches on
# neither side.
CITATION_MARKER = re.compile(r"\[([0-9]{1,2})\]")


def _to_float(value: Any) -> float | None:
    """Coerce numpy/pandas scalars to a plain float, mapping NaN to None."""
    if value is None:
        return None
    try:
        result = float(value)
    except (TypeError, ValueError):
        return None
    return None if math.isnan(result) or math.isinf(result) else round(result, 4)


def _to_int(value: Any) -> int | None:
    if value is None:
        return None
    try:
        result = float(value)
    except (TypeError, ValueError):
        return None
    if math.isnan(result) or math.isinf(result):
        return None
    return int(result)


def _to_str(value: Any, fallback: str = "") -> str:
    if value is None:
        return fallback
    try:
        if isinstance(value, float) and math.isnan(value):
            return fallback
    except (TypeError, ValueError):
        pass
    text = str(value).strip()
    return text or fallback


def _snippet(text: Any) -> str:
    """First SNIPPET_CHARS of the retrieved chunk, whitespace collapsed."""
    collapsed = " ".join(_to_str(text).split())
    if len(collapsed) <= SNIPPET_CHARS:
        return collapsed
    return collapsed[:SNIPPET_CHARS].rsplit(" ", 1)[0] + "…"


def build_source_payload(results: Iterable[Any], limit: int | None = None) -> list[dict[str, Any]]:
    """Serialise SearchResults into JSON-native citation cards.

    ``index`` is 1-based and matches the ``[n]`` label in the prompt context,
    so a marker in the answer resolves to exactly one card.
    """
    payload: list[dict[str, Any]] = []

    for position, result in enumerate(results, start=1):
        if limit is not None and position > limit:
            break

        metadata = getattr(result, "metadata", None) or {}

        payload.append(
            {
                "index": position,
                "document": _to_str(getattr(result, "document", None), "Unknown document"),
                "page": _to_int(getattr(result, "page", None)),
                "category": _to_str(getattr(result, "category", None), "uncategorised"),
                "score": _to_float(getattr(result, "score", None)),
                "semantic_score": _to_float(metadata.get("semantic_score")),
                "lexical_score": _to_float(metadata.get("lexical_score")),
                "chunk_id": _to_str(getattr(result, "chunk_id", None)) or None,
                "snippet": _snippet(getattr(result, "text", None)),
            }
        )

    return payload


def normalize_legacy_citations(text: str, sources: list[dict[str, Any]]) -> str:
    """Rewrite prose ``[Source: Doc, Page: N]`` citations into ``[n]`` markers.

    The model reverts to the old format occasionally — especially mid-migration,
    when the conversation history is full of examples of it. Matching on
    (document, page) is exact; a citation naming a document that is not in the
    retrieved set is dropped rather than mapped to the wrong card.

    The document-only fallback applies ONLY when the model supplied no page at
    all. It used to be an unconditional ``or``, so "[Source: Guidance.pdf,
    Page: 99]" against a retrieved Guidance.pdf p.14 silently produced a marker
    pointing at page 14 — a citation the reader would check, find plausible,
    and never learn was for a different page. A wrong page is now dropped.
    """
    if not text or not sources:
        return text

    by_doc_page: dict[tuple[str, int | None], int] = {}
    by_doc: dict[str, int] = {}
    for source in sources:
        document = source["document"].casefold()
        by_doc_page.setdefault((document, source["page"]), source["index"])
        by_doc.setdefault(document, source["index"])

    dropped = 0

    def replace(match: re.Match[str]) -> str:
        nonlocal dropped
        document = match.group("doc").strip().casefold()
        page = int(match.group("page")) if match.group("page") else None

        if page is None:
            index = by_doc_page.get((document, None)) or by_doc.get(document)
        else:
            index = by_doc_page.get((document, page))

        if index is None:
            dropped += 1
            # The captured leading space goes with it, so nothing downstream
            # has to tidy up after this.
            return ""
        return f"{match.group('lead')}[{index}]"

    result = LEGACY_CITATION.sub(replace, text)

    if dropped:
        logger.info("Dropped %d citation(s) naming documents outside the retrieved set.", dropped)

    # No global whitespace pass. There used to be one here — collapse every run
    # of two-or-more spaces to one, then close the gap before punctuation — to
    # tidy up after dropped citations. It ran over the WHOLE answer on every
    # turn, whether or not anything had been dropped, and markdown is
    # whitespace-sensitive: a four-space-indented code block became an ordinary
    # paragraph, and nested list indentation flattened.
    #
    # For citations specifically that was self-defeating. Indent this:
    #
    #     rows = data[1]
    #
    # and the collapse turns it into prose, so "[1]" stops being code and
    # starts being a citation — on BOTH sides, since the browser is reading the
    # markdown this function already rewrote. The intersection in
    # bindCitations cannot help, because the damage happened before rendering.
    #
    # The tidying is local now: LEGACY_CITATION captures the space in front of
    # a citation and a dropped one takes it along.
    return result


# Regions where a "[1]" is not a citation. `bindCitations` in citations.js
# refuses to linkify inside `pre, code, a`, so a marker counted here that the
# browser will not turn into a button is a source the reader cannot reach.
#
# This is a best-effort approximation and cannot be anything else: these are
# regexes over markdown SOURCE, while the browser walks the tree `marked`
# actually produced. Indented code blocks, unterminated fences, multi-line
# spans, nested link labels and raw HTML all diverge here, and chasing them
# case by case is how "[1][2]" — the multi-citation form the system prompt
# asks for — briefly got read as reference-link syntax and dropped.
#
# The divergence is made harmless downstream rather than eliminated here:
# bindCitations reports which markers it actually bound and the client shows
# only those, so anything this filter misses costs a passage its place in the
# panel, never the reader a source they cannot check. Keep this filter as an
# accuracy improvement for the `cited` telemetry, not as a correctness
# guarantee — that lives on the client.
#
# Fenced blocks and inline spans cover `pre`/`code`; inline link and image
# syntax covers `a`. These are the forms `marked` reliably converts.
#
# Reference-style links ("[label][1]") are deliberately NOT stripped. `marked`
# only builds one when a matching "[1]: url" definition exists, and a model
# answer never contains link definitions — so the text stays literal and
# bindCitations linkifies the marker. Stripping it would create the mismatch
# this function exists to prevent, in the other direction. It also collided
# with "[1][2]", the multi-citation form BASE_SYSTEM_MESSAGE explicitly asks
# for, silently dropping both markers.
_UNCITABLE = re.compile(
    r"```.*?```"  # fenced code block
    r"|~~~.*?~~~"  # the other fence
    r"|`[^`\n]*`"  # inline code span
    r"|!?\[[^\]\n]*\]\([^)\n]*\)",  # [text](url) / ![alt](src)
    re.S,
)


# Alternation order is the mechanism: a code or link region matches FIRST and
# is handed back untouched, so a marker inside one is never reached by the
# `marker` branch.
_MARKER_OUTSIDE_CODE = re.compile(
    _UNCITABLE.pattern + r"|(?P<marker> ?\[[0-9]{1,2}\])",
    re.S,
)


def strip_citation_markers(text: str) -> str:
    """Remove "[n]" markers, for text that is about to outlive its numbering.

    A citation number means nothing on its own — it is an index into the
    context blocks of ONE request. Every request numbers its own retrieved
    passages from [1], so a marker that survives into a later turn points at
    whatever that later turn happened to retrieve first.

    That is not hypothetical: completed answers go into the conversation
    history and are replayed to the model as prior context. An answer saying
    "submit within 15 days [1]" comes back on turn two, where [1] is now a
    different passage from a different document. A model that restates the
    claim carries the marker with it, and the answer ships a citation
    resolving to unrelated evidence — with a working link and a real passage
    behind it, which is exactly what makes it dangerous.

    Takes any single space in front of the marker with it, so "days [1]."
    becomes "days." rather than "days ." — the same local rule
    ``normalize_legacy_citations`` uses, and for the same reason: no pass over
    the surrounding text, because indentation is markdown.

    Code and links are left verbatim. An answer containing ``data[1]`` in a
    fenced block is replayed to the model as prior context, and quietly
    rewriting it to ``data`` would teach the model to produce broken code.
    Indented code blocks are NOT protected: four spaces inside a list item is
    continuation rather than code, and this cannot tell the two apart without a
    parser. That direction is the safe one to get wrong here — a stray marker
    surviving into history is inert, whereas mangling a list would not be.
    """
    if not text:
        return text
    return _MARKER_OUTSIDE_CODE.sub(
        lambda match: "" if match.group("marker") else match.group(0), text
    )


def _strip_uncitable(text: str) -> str:
    """Blank out code and link regions, preserving offsets elsewhere."""
    return _UNCITABLE.sub(lambda m: " " * len(m.group(0)), text)


@dataclass(frozen=True)
class CitationDiagnostics:
    """Every ``[n]`` marker in an answer, classified.

    ``extract_cited_indices`` used to compute this same breakdown and then
    throw away everything but ``cited`` — ``invalid`` was already being
    counted, just for a single ``logger.info`` line (see below) with no
    caller able to turn it into a rate. This is that breakdown made
    returnable, so a citation-fidelity harness can aggregate it instead of
    scraping log lines.

    Attributes:
        cited: Indices present in ``sources`` that the text actually cited —
            identical to what ``extract_cited_indices`` returns.
        invalid: Indices the text cited that are NOT in ``sources`` —
            hallucinated or prompt-drifted markers, in encounter order with
            duplicates kept (so a rate computed over ``total_markers``
            reflects how often this happens, not just which values occurred).
        total_markers: Every ``[n]`` matched, valid or not — the denominator
            for a hallucinated-marker rate.
    """

    cited: list[int]
    invalid: list[int]
    total_markers: int


def extract_citation_diagnostics(text: str, sources: list[dict[str, Any]]) -> CitationDiagnostics:
    """Classify every ``[n]`` marker in *text* against *sources*.

    The single parsing pass behind both this function and
    ``extract_cited_indices`` — see that function's docstring for the
    membership-vs-range rule, the refusal-detection rationale, and what this
    is and is not evidence of. ``extract_cited_indices`` is a thin wrapper
    around this that returns only ``.cited``, kept for its existing callers
    and the citation-panel contract; anything that also wants the
    hallucinated-marker count should call this directly rather than
    re-deriving it from a second regex pass (as ``scripts/smoke_real.py`` used
    to).

    Args:
        text: The answer, AFTER ``normalize_legacy_citations`` — the same
            string the reader is shown.
        sources: The source payload, as built by ``build_source_payload``.

    Returns:
        A ``CitationDiagnostics`` with empty/zero fields when *text* or
        *sources* is empty.
    """
    if not text or not sources:
        return CitationDiagnostics(cited=[], invalid=[], total_markers=0)

    valid = {source["index"] for source in sources}
    cited: set[int] = set()
    invalid: list[int] = []
    total = 0

    for match in CITATION_MARKER.finditer(_strip_uncitable(text)):
        index = int(match.group(1))
        total += 1
        if index in valid:
            cited.add(index)
        else:
            invalid.append(index)

    if invalid:
        # Prompt drift and hallucinated citations both surface here. This is
        # the telemetry behind the decision to keep showing retrieval
        # candidates when an answer cites nothing.
        logger.info(
            "Ignored %d citation marker(s) outside the source set %s: %s",
            len(invalid),
            sorted(valid),
            sorted(set(invalid)),
        )

    return CitationDiagnostics(cited=sorted(cited), invalid=invalid, total_markers=total)


def extract_cited_indices(text: str, sources: list[dict[str, Any]]) -> list[int]:
    """The source indices *text* actually cites, sorted and deduplicated.

    This is the signal that decides whether an answer gets a source panel. It
    replaces the previous arrangement, where sources were emitted before the
    model was even called and so could not reflect what the answer did with
    them — a refusal shipped with eight source cards attached.

    Deliberately NOT a check for refusal wording. The refusal is an
    instruction the model paraphrases (``BASE_SYSTEM_MESSAGE``), and no Arabic
    refusal string exists anywhere — the Arabic wording is entirely
    model-generated. Counting markers is language-agnostic and needs no list
    of phrases to keep current.

    Validated against the ``index`` values actually present in *sources*, not
    against ``1..len(sources)``: once the payload is filtered to cited
    passages the indices are sparse (1, 3, 7), and a range check would then
    accept ``[2]`` for a passage that is not in the payload.

    A marker outside that set is dropped, matching ``bindCitations`` in
    ``static/js/modules/citations.js``, which leaves an out-of-range marker as
    literal text rather than linking it to nothing. A model citing ``[9]``
    against 8 sources is hallucinating, and counting that as evidence would be
    worse than counting nothing.

    Note what this does and does not prove: that the model *emitted* ``[2]``,
    not that passage 2 supports the claim. It is a citation signal, not a
    grounding one.

    Args:
        text: The answer, AFTER ``normalize_legacy_citations`` — the same
            string the reader is shown. Running this on pre-normalization text
            reports citations the browser never rendered.
        sources: The source payload, as built by ``build_source_payload``.

    Returns:
        Sorted unique indices, or ``[]`` when nothing valid was cited.
    """
    return extract_citation_diagnostics(text, sources).cited

File: static/js/modules/citations.js
/**
 * SFDA Copilot — Citations
 *
 * The answer text arrives carrying numbered markers ([1], [2][5]) that the
 * model wrote against numbered context blocks. This module turns those into
 * interactive markers, and renders the one-line trigger that opens the
 * passages behind them.
 *
 * The passages themselves live in source-panel.js. They used to be rendered
 * here, as a deck of cards under the answer; see that module's header for why
 * they moved.
 *
 * Load-bearing detail: markers are built with createElement AFTER the markdown
 * has been sanitized, so DOMPurify never sees them. Never string-replace into
 * the HTML before sanitizing — that would put attacker-influenced answer text
 * back through the parser.
 */

import { DOMCache } from './dom.js';
import { iconMarkup } from './icons.js';
import { I18n } from './i18n.js';
import { SourcePanel, groupByDocument, isDatedEvidence } from './source-panel.js';

/* Two regexes, deliberately. Sharing one /g regex between .test() in
   acceptNode and .exec() in the replace loop corrupts lastIndex and silently
   drops every other match. */
const CITE_TEST = /\[\d{1,2}\]/;
const CITE_SCAN = /\[(\d{1,2})\]/g;

/**
 * A message id that cannot collide with one from an earlier page load.
 *
 * This was a counter — m1, m2, m3 — reset to zero on every load. The transcript
 * survives a language switch by being re-inserted as saved HTML, so a restored
 * answer arrived still carrying `data-msg="m1"` while the counter handed the
 * NEXT answer that same id. Clicking the old answer's [1] then opened the new
 * answer's sources: a citation resolving to evidence from a different question,
 * silently, with nothing on screen to suggest anything was wrong. On a
 * regulatory surface that is the worst failure this module can produce, worse
 * than showing no sources at all.
 *
 * randomUUID needs a secure context, which http:// on a LAN address is not, so
 * the fallback keeps the guarantee where it is unavailable.
 */
export function nextMessageId() {
  if (crypto?.randomUUID) return `m${crypto.randomUUID()}`;
  return `m${Date.now().toString(36)}${Math.random().toString(36).slice(2, 10)}`;
}

/* Every answer's source state, keyed by message id. The panel is global and
   shows one answer at a time, so it cannot hold this itself — a reader
   clicking a marker in an older answer has to be able to swap the panel back
   to that answer's passages.

   Bounded because it is otherwise a slow leak: each entry holds up to eight
   passages with their snippets, nothing removed an entry, and a long session
   on one tab accumulates all of them. The cap is generous — far beyond the
   transcript anyone scrolls back through — and eviction is oldest-first,
   which matches the direction people read. An evicted answer's controls stop
   resolving rather than resolving wrongly: the click handler bails when the
   id is unknown. */
const MAX_TRACKED_ANSWERS = 100;
const stateByMessage = new Map();

function rememberSources(msgId, state) {
  stateByMessage.set(msgId, state);
  while (stateByMessage.size > MAX_TRACKED_ANSWERS) {
    // Map iterates in insertion order, so this is the oldest answer.
    stateByMessage.delete(stateByMessage.keys().next().value);
  }
}

function makeMarker(index, msgId) {
  const button = DOMCache.createElement('button', 'cite-marker');
  button.type = 'button';
  button.dataset.cite = String(index);
  button.dataset.msg = msgId;
  button.textContent = String(index);
  button.setAttribute('aria-label', I18n.t('cite.marker', { n: index }));
  button.setAttribute('aria-controls', 'source-panel');
  return button;
}

/**
 * Replace [n] text with marker elements throughout `scope`.
 *
 * Returns the set of indices that actually became buttons. That return value
 * is what makes this function the authority on which sources an answer can
 * display: the server decides which indices are ALLOWED (membership), the
 * browser decides which are REACHABLE, and the panel shows the intersection.
 *
 * Without it the two sides had to agree about markdown, and they cannot. The
 * server matches regexes against the raw source; this walks the tree `marked`
 * actually produced. Every attempt to close that gap case by case — fenced
 * blocks, inline spans, link syntax — is a guess about a parser, and one such
 * guess already misread "[1][2]" as reference-link syntax and silently dropped
 * a sentence's entire provenance. Intersecting the two answers instead makes
 * the whole class safe by construction: an index the browser could not bind is
 * one the reader could not click, so it is not offered as a source.
 *
 * An index the payload does not carry is left as literal text rather than
 * linkified — a model that cites [9] against 8 sources is hallucinating, and
 * silently linking it to nothing would hide that.
 *
 * Membership, not range: once the payload is filtered to the passages the
 * answer cited, the indices are sparse (1, 3, 7). A `n <= sources.length`
 * test would accept [2] and point it at nothing. This mirrors
 * `extract_cited_indices` in web/services/citations.py, which validates the
 * same way — the two must agree or the server reports a citation the reader
 * cannot click.
 */
export function bindCitations(scope, sources, msgId) {
  const bound = new Set();
  if (!scope || !Array.isArray(sources) || !sources.length) return bound;
  const valid = new Set(sources.map((s) => s.index));

  const walker = document.createTreeWalker(scope, NodeFilter.SHOW_TEXT, {
    acceptNode(node) {
      // Never rewrite inside code, existing links, or a marker we already made.
      if (node.parentElement?.closest('pre, code, a, .cite-marker')) {
        return NodeFilter.FILTER_REJECT;
      }
      return CITE_TEST.test(node.nodeValue) ? NodeFilter.FILTER_ACCEPT : NodeFilter.FILTER_REJECT;
    },
  });

  // Collect before mutating: replacing nodes mid-walk invalidates the walker.
  const nodes = [];
  for (let node; (node = walker.nextNode());) nodes.push(node);

  for (const node of nodes) {
    const text = node.nodeValue;
    const fragment = document.createDocumentFragment();
    let last = 0;
    let matched = false;
    let match;

    CITE_SCAN.lastIndex = 0;
    while ((match = CITE_SCAN.exec(text))) {
      const index = Number(match[1]);
      if (!valid.has(index)) continue;

      fragment.append(text.slice(last, match.index), makeMarker(index, msgId));
      bound.add(index);
      last = match.index + match[0].length;
      matched = true;
    }

    if (!matched) continue;
    fragment.append(text.slice(last));
    node.replaceWith(fragment);
  }

  return bound;
}

/**
 * The one line an answer's sources get in the reading column.
 *
 * `state.sources` is what the answer CITED **and** the reader can reach —
 * callers pass the set already intersected with what bindCitations bound. An
 * answer that cited nothing gets an empty list from the server and therefore
 * no element here at all, which is the whole fix. A refusal reading "I cannot
 * answer based on the given information" must not be followed by any source
 * control: the evidence the surface is offering belongs to no claim the answer
 * made.
 *
 * There was an intermediate design where a citation-free answer kept a muted
 * "8 passages retrieved, not cited" line, so a reader auditing the answer
 * could still see what search returned. It reads as a contradiction — the
 * label disclaims the passages while advertising eight of them — and under a
 * refusal it is still evidence attached to an answer that has none. Retrieval
 * candidates are a server-side diagnostic now.
 *
 * Returns null when there is nothing to show, so callers can skip appending.
 */
export function renderSourceTrigger(state, msgId) {
  const sources = state?.sources || [];
  if (!sources.length) return null;

  rememberSources(msgId, state);

  const button = DOMCache.createElement('button', 'source-trigger');
  button.type = 'button';
  button.dataset.msg = msgId;
  button.setAttribute('aria-controls', 'source-panel');
  button.setAttribute('aria-expanded', 'false');

  // Both numbers count the same (cited) set, so they cannot contradict.
  const documents = groupByDocument(sources).length;
  const label =
    `${I18n.t('cite.sourcesLabel')} · ` +
    `${I18n.plural(documents, 'cite.docsOne', 'cite.docsMany')} · ` +
    `${I18n.plural(sources.length, 'cite.passagesOne', 'cite.passagesMany')}`;

  const text = DOMCache.createElement('span');
  text.textContent = label;
  button.append(text);

  /* A stored answer whose corpus has been rebuilt under it still opens — the
     row behind it holds the document, page and passage the model actually
     read, frozen when the answer was written, so opening it shows what was
     read rather than a guess about where that text lives now.

     What changes is that the reader is TOLD. Silently serving a passage from a
     superseded build as though it were current is the one failure this product
     cannot afford; silently withholding it is barely better, because the
     reader cannot then tell "this evidence is dated" from "this answer had no
     sources". `stale` and `unverifiable` are one badge on purpose: to a reader
     they mean the same thing — we cannot confirm this is still in the live
     corpus — and they stay distinct in the payload for logs and tests. */
  if (isDatedEvidence(state?.evidenceState)) {
    button.dataset.evidence = state.evidenceState;
    const badge = DOMCache.createElement('span', 'source-trigger-badge');
    badge.textContent = I18n.t('cite.datedBadge');
    button.append(badge);
  }

  button.insertAdjacentHTML('beforeend', iconMarkup('chevron-right', 12, 'chev'));
  return button;
}

/** Forget every answer's sources, e.g. on logout. */
export function resetCitationState() {
  stateByMessage.clear();
}

/* `neutraliseRestoredCitations` lived here until step 6 and is deliberately
   gone rather than merely unused.
 *
 * It stripped the source controls from a transcript restored out of
 * `sessionStorage`, because only the MARKUP came back — `stateByMessage` lives
 * in module memory, so the passages behind a restored answer were genuinely
 * lost and its controls would have opened an empty panel. A control that does
 * nothing is a worse lie than no control, so the markers reverted to plain
 * text.
 *
 * The transcript now hydrates from durable rows through the same render path a
 * live answer takes, so a restored answer arrives WITH its passages and its
 * controls resolve. Keeping the function would leave a loaded gun: called on a
 * hydrated transcript it would silently strip evidence that is present and
 * correct. */

/**
 * One delegated listener set for the whole transcript, so markers created
 * during streaming are live without rebinding.
 */
export function initCitationInteractions(container) {
  if (!container || container.dataset.citationsBound) return;
  container.dataset.citationsBound = '1';
  SourcePanel.init();

  const highlight = (event) => {
    const marker = event.target.closest?.('.cite-marker');
    if (marker) SourcePanel.highlight(marker.dataset.msg, Number(marker.dataset.cite));
  };
  const clear = (event) => {
    if (event.target.closest?.('.cite-marker')) {
      SourcePanel.highlight(event.target.closest('.cite-marker').dataset.msg, null);
    }
  };

  container.addEventListener('pointerover', highlight);
  container.addEventListener('focusin', highlight);
  container.addEventListener('pointerout', clear);
  container.addEventListener('focusout', clear);

  container.addEventListener('click', (event) => {
    const marker = event.target.closest('.cite-marker');
    if (marker) {
      const msgId = marker.dataset.msg;
      const state = stateByMessage.get(msgId);
      if (!state) return;
      SourcePanel.open(msgId, state, {
        focusIndex: Number(marker.dataset.cite),
        returnFocus: marker,
      });
      return;
    }

    const trigger = event.target.closest('.source-trigger');
    if (trigger) {
      const msgId = trigger.dataset.msg;
      // A second press on the trigger that opened it closes it again.
      // aria-expanded is not touched here: SourcePanel owns it, so that every
      // route in and out leaves exactly one trigger claiming to be open.
      if (SourcePanel.isOpenFor(msgId)) {
        SourcePanel.close();
        return;
      }
      const state = stateByMessage.get(msgId);
      if (!state) return;
      SourcePanel.open(msgId, state, { returnFocus: trigger });
    }
  });
}
